aarch32: use vector base address register for v7-A

yroeht opened the pull request:
Do not spuriously copy the vector to zero unless there are no means to set VBAR.

Signed-off-by: Théophile Ranquet <theophile.ranquet@gmail.com>

Fixes #51024

yroeht committed:
aarch32: mmu: include exception vector into default range

This removes the need to include an entry in the soc's
mmu_config mmu_regions, in some cases.

Signed-off-by: Théophile Ranquet <theophile.ranquet@gmail.com>

yroeht committed:
aarch32: use vector base address register for v7-A

Do not spuriously copy the vector to zero unless there are no means to
set VBAR.

Signed-off-by: Théophile Ranquet <theophile.ranquet@gmail.com>

yroeht commented:
I believe this #elif could be expanded to included CONFIG_AARCH32_ARMV7_R but I am no expert. I think the #else is only for the CORTEX M which do not define CONFIG_CPU_CORTEX_M_HAS_VTOR (Cortex-M0 I think?)

ibirnbaum commented:
At the time being, two SoCs actively zero VBAR: the Zynq-7000/7000S (added by me), and the Intel Cyclone V.

In my case, the motivation for doing so was that for a long time, my Zephyr builds always got loaded via JTAG where all registers start with their default power-on values. When eventually transitioning to booting via u-boot, it turned out that u-boot uses VBAR for vectors positioning and doesn't clean up after itself when handing over to the OS (in general, not board-specific behaviour. There's a single B&R board for which VBAR gets reset because otherwise VxWorks won't boot on that board). So I had to revert to a world where the vectors are either located at address zero or the HIVECS address.
The zeroing of VBAR on the Zynq is executed in z_arm_platform_init, which is about as early as it gets, definitely before vector relocation. Therefore, this doesn't collide with your modification.

On the Cyclone however, VBAR is zeroed in soc_intel_cyclonev_init, which is registered in the PRE_KERNEL_1 phase - IIRC, that's *after* the vector relocation. Could you check up on that? If your modification takes place first, you should preferrably amend this PR to prevent the Cyclone V from crashing upon boot time.

ibirnbaum commented:
Speaking of HIVECS: not that there's currently a Kconfig option for it, but maybe it should be checked who overrides whom if we had one: if HIVECS was set in SCTLR, would that override VBAR or not? Preferrably it should be kept in mind that someday someone might want to determine the location of the vectors purely based on the HIVECS flag, which should then take precedence, resulting in the vectors being copied to that location rather than just having VBAR point to the location of the vectors in the Zephyr image in RAM.

yroeht commented:
I think samples/compression/lz4/sample.compression.lz4 is failing due to https://github.com/zephyrproject-rtos/zephyr/issues/51024

Reason unknown for the others:

tests/kernel/mem_protect/mem_map/
tests/kernel/common/kernel.common.picolibc 
samples/subsys/logging/syst/sample.logger.syst.immediate
samples/subsys/logging/syst/sample.logger.syst.deferred
samples/subsys/logging/syst/sample.logger.syst.immediate_cpp
samples/subsys/logging/syst/sample.logger.syst.catalog.immediate_cpp
samples/subsys/logging/syst/sample.logger.syst.catalog.immediate
etc.

ibirnbaum commented:
> I think samples/compression/lz4/sample.compression.lz4 is failing due to #51024
> 
> Reason unknown for the others:
> 
> tests/kernel/mem_protect/mem_map/ tests/kernel/common/kernel.common.picolibc samples/subsys/logging/syst/sample.logger.syst.immediate samples/subsys/logging/syst/sample.logger.syst.deferred samples/subsys/logging/syst/sample.logger.syst.immediate_cpp samples/subsys/logging/syst/sample.logger.syst.catalog.immediate_cpp samples/subsys/logging/syst/sample.logger.syst.catalog.immediate etc.

This may or may not be related, but as the majority of the mentioned test cases are from the logging subsystem, this does ring a bell for me. At least one of the logging test cases does stuff with data in memory (from .rodata, IIRC), and checks against the corresponding section start/end markers generated via the linker command file. When MMU support was introduced, one of those tests failed because of different behaviour of the '.' operator in the linker command file, depending on whether it was part of a group or not. It's a tad difficult to explain in text form, you might want to have a look at #41513.

Are you running your tests in conjunction with the changes you made in #51024? If so, maybe your move of z_mapped_start causes an issue for tests checking against such markers if its location changed from within a group to outside of a group.

yroeht committed:
aarch32: pin excn vector in mmu

As discussed in #51029, the vector not being included in mapped memory
causes it to be overwritten by paging mecanisms.

Signed-off-by: Théophile Ranquet <theophile.ranquet@gmail.com>

yroeht commented:
Ok so I was right that some failures were due to not implementing https://github.com/zephyrproject-rtos/zephyr/issues/51024 , so I have included that as a second commit in this merge request.

I found this while experimenting with

`./zephyr/scripts/twister -s tests/kernel/mem_protect/mem_map/kernel.memory_protection.mem_map -p qemu_cortex_a9`

so it might not address all the failures. Wait and see...

yroeht commented:
FYI current test results

```
INFO    - Total complete:  811/1257  64%  skipped:  729, failed:    0
ERROR   - qemu_cortex_a9            tests/kernel/timer/timer_api/kernel.timer.tickless FAILED: Timeout
ERROR   - see: /home/tranquet/zephyrproject-yroeht/twister-out/qemu_cortex_a9/tests/kernel/timer/timer_api/kernel.timer.tickless/handler.log                                                                      
INFO    - Total complete:  812/1257  64%  skipped:  729, failed:    1
ERROR   - qemu_cortex_a9            tests/kernel/timer/timer_api/kernel.timer          FAILED: Timeout
ERROR   - see: /home/tranquet/zephyrproject-yroeht/twister-out/qemu_cortex_a9/tests/kernel/timer/timer_api/kernel.timer/handler.log                                                                               
INFO    - Total complete:  840/1257  66%  skipped:  729, failed:    2
ERROR   - qemu_cortex_a9            tests/net/socket/tcp/net.socket.tcp                FAILED: Failed
ERROR   - see: /home/tranquet/zephyrproject-yroeht/twister-out/qemu_cortex_a9/tests/net/socket/tcp/net.socket.tcp/handler.log                                                                                     
INFO    - Total complete: 1158/1257  92%  skipped:  820, failed:    3
ERROR   - qemu_cortex_a9            tests/subsys/logging/log_api/logging.log_api_deferred_override_level.tagged_args FAILED: Timeout                                                                              
ERROR   - see: /home/tranquet/zephyrproject-yroeht/twister-out/qemu_cortex_a9/tests/subsys/logging/log_api/logging.log_api_deferred_override_level.tagged_args/build.log                                          
INFO    - Total complete: 1159/1257  92%  skipped:  820, failed:    4
ERROR   - qemu_cortex_a9            tests/subsys/logging/log_api/logging.log_api_deferred_override_level_rt_filtering.tagged_args FAILED: Timeout                                                                 
ERROR   - see: /home/tranquet/zephyrproject-yroeht/twister-out/qemu_cortex_a9/tests/subsys/logging/log_api/logging.log_api_deferred_override_level_rt_filtering.tagged_args/build.log                             
INFO    - Total complete: 1188/1257  94%  skipped:  820, failed:    5
ERROR   - qemu_cortex_a9            tests/subsys/logging/log_api/logging.log_api_frontend_immediate_override_level FAILED: Timeout                                                                                
ERROR   - see: /home/tranquet/zephyrproject-yroeht/twister-out/qemu_cortex_a9/tests/subsys/logging/log_api/logging.log_api_frontend_immediate_override_level/build.log                                            
INFO    - Total complete: 1191/1257  94%  skipped:  820, failed:    6
ERROR   - qemu_cortex_a9            tests/subsys/logging/log_api/logging.log_api_deferred_override_level_rt_filtering FAILED: Timeout                                                                             
ERROR   - see: /home/tranquet/zephyrproject-yroeht/twister-out/qemu_cortex_a9/tests/subsys/logging/log_api/logging.log_api_deferred_override_level_rt_filtering/build.log                                         
INFO    - Total complete: 1194/1257  94%  skipped:  820, failed:    7
ERROR   - qemu_cortex_a9            tests/subsys/logging/log_api/logging.log_api_deferred_override_level FAILED: Timeout                                                                                          
ERROR   - see: /home/tranquet/zephyrproject-yroeht/twister-out/qemu_cortex_a9/tests/subsys/logging/log_api/logging.log_api_deferred_override_level/build.log                                                      
INFO    - Total complete: 1257/1257  100%  skipped:  839, failed:    8
INFO    - 1371 test scenarios (1257 test instances) selected, 839 configurations skipped (673 by static filter, 166 at runtime).                                                                                  
INFO    - 410 of 1257 test configurations passed (98.09%), 8 failed, 839 skipped with 0 warnings in 2003.83 seconds                                                                                               
INFO    - In total 3495 test cases were executed, 4169 skipped on 1 out of total 505 platforms (0.20%)
INFO    - 375 test configurations executed on platforms, 43 test configurations were only built.

```

yroeht commented:
> > I think samples/compression/lz4/sample.compression.lz4 is failing due to #51024
> > Reason unknown for the others:
> > tests/kernel/mem_protect/mem_map/ tests/kernel/common/kernel.common.picolibc samples/subsys/logging/syst/sample.logger.syst.immediate samples/subsys/logging/syst/sample.logger.syst.deferred samples/subsys/logging/syst/sample.logger.syst.immediate_cpp samples/subsys/logging/syst/sample.logger.syst.catalog.immediate_cpp samples/subsys/logging/syst/sample.logger.syst.catalog.immediate etc.
> 
> This may or may not be related, but as the majority of the mentioned test cases are from the logging subsystem, this does ring a bell for me. At least one of the logging test cases does stuff with data in memory (from .rodata, IIRC), and checks against the corresponding section start/end markers generated via the linker command file. When MMU support was introduced, one of those tests failed because of different behaviour of the '.' operator in the linker command file, depending on whether it was part of a group or not. It's a tad difficult to explain in text form, you might want to have a look at #41513.
> 
> Are you running your tests in conjunction with the changes you made in #51024? If so, maybe your move of z_mapped_start, which is based on the '.' operator, causes an issue for tests checking against such markers if its location changed from within a group to outside of a group or is transferred from one group to another.

Interesting, but as far as I can tell the location of z_mapped_start has not changed to outside a group or another group.

I generated the elf with `
west build --pristine -b qemu_cortex_a9 zephyr/tests/subsys/logging/log_api/` for both before my commit (commit 25f14d7092) and after, here is what I got:

```
diff --git a/25f14d7092.map b/new.map
index 487ce6a..d01ef8a 100644
--- a/25f14d7092.map
+++ b/new.map
@@ -1673,6 +1673,7 @@ LOAD zephyr/CMakeFiles/offsets.dir/./arch/arm/core/offsets/offsets.c.obj
                 0x0000000000100000                __rom_region_start = 0x100000
 
 rom_start       0x0000000000100000       0x3c
+                0x0000000000100000                z_mapped_start = .
                 0x0000000000000000                . = 0x0
                 0x0000000000000000                . = ALIGN (0x4)
                 0x0000000000100000                _vector_start = .
@@ -1688,7 +1689,6 @@ text            0x0000000000100040     0xbfc0
                 0x0000000000101000                . = ALIGN (_region_min_align)
  *fill*         0x0000000000100040      0xfc0 
                 0x0000000000101000                __text_region_start = .
-                0x0000000000101000                z_mapped_start = .
- 
```

And both have identical rodata_start

```
rodata_start    0x000000000010c008      0xff8                                                                                                                                                                      
                0x000000000010d000                . = ALIGN (_region_min_align)                                                                                                                                    
 *fill*         0x000000000010c008      0xff8.                                                                                                                                                                     
                0x000000000010d000                __rodata_region_start = .                                                                                                                                        
                                                                                                                                                                                                                   
initlevel       0x000000000010d000       0x38                                                                                                                                                                      
                0x000000000010d000                __init_start = .        
```

So it sounded promising but I am not sure this is the problem.

yroeht commented:
Test logging.log_api_frontend_immediate_override_level (and others presumably) fails, and I have noted the following conditions:

The executable zephyr/tests/subsys/logging/log_api/
  - runs OK when given option CONFIG_LOG_OVERRIDE_LEVEL=0
  - fails when given option CONFIG_LOG_OVERRIDE_LEVEL=1 or 2 or 3
  - hangs when given option CONFIG_LOG_OVERRIDE_LEVEL=4, with the following being output: ```
$ west build -t run
-- west build: running target run
[0/1] To exit from QEMU enter: 'CTRL+a, x'[QEMU] CPU: cortex-a9
qemu-system-xilinx-aarch64: warning: nic ethernet@e000b000 has no peer
qemu-system-xilinx-aarch64: warning: nic ethernet@e000c000 has no peer
```

ibirnbaum commented:
Just before the weekend I tested your PR, although in a slightly different manner: instead of cloning your forked repo and switching to your main2 branch, I applied the hand full of changes contained in this PR manually to what was then my current main.

Doing it that way, I couldn't find any issues - neither on real hardware, nor on qemu_cortex_a9. I tested the lz4 compression sample and tests/kernel/mem_protect/mem_map, For the latter, I looked at the MMU table dump that the Lauterbach debugger provides, and it all made sense. I also picked samples/subsys/logging/syst/sample.logger.syst.immediate out of the logging testcases you mentioned - no issues here either. This morning, I cloned your main2 and I'm currently having a look at that. I didn't look at the specific test cases you mentioned yet, but rather focused on whether or not VBAR works in QEMU. Using the Xilinx QEMU that comes with the SDK 0.15.1, the answer seems to be yes - I built some of the simpler synchronization samples which invoke the scheudler either explicitly or via a timer interrupt, and they work. Using memory dumps, I can see that address 0x0 is unmapped, so the vectors definitely aren't there, and the two important handler functions pointed to by the vectors used during regular operation (z_arm_svc for SW-based scheduler invocation and _isr_wrapper) both work fine. Therefore, the vectors seem to work as expected. I wouldn't care too much about your console VBAR output, who knows how and to which degree it is implemented. I'd blame this on VBAR being a CP15 register for which probably no proper GDB-multiarch integration exists, considering that debugger vendors like Segger provide custom CP15 extensions for GDB when targeting their devices. If VBAR didn't work, there wouldn't be a single timer interrupt nor a single context switch.

Now on to the testcases that you've observed failing...

ibirnbaum commented:
Something odd is happening with regards to the log_api test case with override level 4:

The system crashes with a data abort, and there's no dump on the console as this probably happens early on. I haven't checked if this happens before the UART/console init. The fault is caused by an attempt to read address 0x4. This would have worked with the old vectors, but now won't work any more. Who's triggering this read operation? The system timer driver, as it tries to read the current counter register contents.

The timer driver has since the initial commit used the statement `DEVICE_MMIO_TOPLEVEL_STATIC(timer_regs, ARM_TIMER_NODE);` to map the timer's physical address to a virtual address if MMU facilities are available. What happens in any function that reads from or writes to the mapped register space is that the macro `TIMER_REG_GET` is used in order to resolve the virtual base address of the timer registers, the offset of the desired register gets added to that, and this value is then passed on to sys_read32/sys_write32.

At any override level other than 4, looking into sys_read32 upon the first read access to the timer's current counter value shows that the 'macro magic' works as expected:

- the target address is 0x008F.D204
- 0x008F.0000 is an address offset that is valid as a virtual address
- 0x204 is the correct offset within the timer's register space for the Global_Timer_Counter_Register1.
- The register's physical address is 0xF8F0.0204

So this all checks out. However, with the override level set to 4, this whole timer_regs mapping / resolving breaks down. There might be something wrong with the TIMER_REG_GET (or its nested DEVICE_MMIO_TOPLEVEL_GET) macro, the timer_regs struct, whatever it may look like, might have been generated incorrectly or was actively overwritten, ...
The thing is that GDB isn't particularly useful when it comes to gaining insight into what kind of structure timer_regs actually has and what its contents are. Also, detecting corruption in case of an accidental overwrite can't be diagnosed that well using QEMU either. I hope that I'll be able to reproduce the issue on my actual Zynq hardware, in that case, the Lauterbach debugger would be a way more powerful tool for finding out what's going wrong.

The individual failed testcases at override levels 1-3 are likely not related to this. Somehow setting override level 0 causes chaos either in memory or in code. I hope that I'll get to the bottom of this in hardware. I haven't tried this test case at level 0 on actual hardware yet, so there's still a chance that it will also fail outside of QEMU.

yroeht commented:
From what I can tell

```
(gdb) l
352     void *memset(void *buf, int c, size_t n)
353     {
354             /* do byte-sized initialization until word-aligned or finished */
355
356             unsigned char *d_byte = (unsigned char *)buf;
357             unsigned char c_byte = (unsigned char)c;
358
359     #if !defined(CONFIG_MINIMAL_LIBC_OPTIMIZE_STRING_FOR_SIZE)
360             while (((uintptr_t)d_byte) & (sizeof(mem_word_t) - 1)) {
361                     if (n == 0) {
(gdb) p buf
$1 = (void *) 0x1d2f28 <z_arm_sys_stack+7848>
(gdb) p /x buf
$2 = 0x1d2f28
(gdb) p n
$3 = 46
(gdb) n
389             while (n > 0) {
(gdb) 
390                     *(d_byte++) = c_byte;
(gdb) 
391                     n--;
(gdb) u
^C
Program received signal SIGINT, Interrupt.
z_mapped_start () at /home/tranquet/zephyrproject-yroeht/zephyr/arch/arm/core/aarch32/cortex_a_r/vector_table.S:24
24              ldr pc, =z_arm_data_abort        /* data abort     offset 0x10 */
```

yroeht commented:
I agree, this is what I have found so far:

```
(gdb) target remote localhost:1234                                                                                                                                                                                
Remote debugging using localhost:1234                                                                                                                                                                             
z_arm_reset () at /home/tranquet/zephyrproject-yroeht/zephyr/arch/arm/core/aarch32/cortex_a_r/reset.S:204                                                                                                         
204         msr CPSR_c, #(MODE_FIQ | I_BIT | F_BIT)
(gdb) b *0x00101d40
Breakpoint 1 at 0x101d40: file /home/tranquet/zephyrproject-yroeht/zephyr/include/zephyr/arch/arm/aarch32/cortex_a_r/timer.h, line 134.                                                                           
(gdb) displ /i $pc                                                                                                                                                                                                
1: x/i $pc                                                                                                                                                                                                        
=> 0x108880 <z_arm_reset>:      msr     CPSR_c, #209    ; 0xd1
(gdb) c                                                                                                                                                                                                           
Continuing.                                                                                                                                                                                                       
                                                                                                                                                                                                                  
Breakpoint 1, 0x00101d40 in arm_arch_timer_count () at /home/tranquet/zephyrproject-yroeht/zephyr/include/zephyr/arch/arm/aarch32/cortex_a_r/timer.h:134                                                          
134             upper = sys_read32(TIMER_REG_GET(TIMER_CNT_UPPER));                                                                                                                                               
1: x/i $pc
=> 0x101d40 <sys_clock_elapsed+88>:     ldr     r2, [r3]
(gdb) si
0x00101d44 in sys_read32 (addr=4) at /home/tranquet/zephyrproject-yroeht/zephyr/include/zephyr/arch/arm/aarch32/cortex_a_r/sys_io.h:57
57              uint32_t val = *(volatile uint32_t *)addr;
1: x/i $pc
=> 0x101d44 <sys_clock_elapsed+92>:     ldr     r1, [r2, #4]
(gdb) p $r2
$1 = 0
(gdb) si
_vector_table () at /home/tranquet/zephyrproject-yroeht/zephyr/arch/arm/core/aarch32/cortex_a_r/vector_table.S:24
24              ldr pc, =z_arm_data_abort        /* data abort     offset 0x10 */
1: x/i $pc
=> 0x100010 <_vector_table+16>: ldr     pc, [pc, #24]   ; 0x100030 <_vector_table+48>
```

ibirnbaum commented:
I think I found the reason for the crash when accessing the timer's registers.

As the init entries of the PRE_KERNEL_1 phase are being processed, their order is:

- soc_xlnx_zynq7000s_init
- arm_gic_init
- syscon_generic_init

And that's the point where things go wrong. Yet, the Syscon init code isn't to blame, the issue rather is that it attempts to map its register space via the MMU. And the MMU's z_phys_map function contains a LOG_DBG statement, which is enforced with override level = 4. The logging module attempts to obtain a timestamp for the debug message, the timer hasn't been initialized yet, therefore the memory access fails:

![syscon-stack](https://user-images.githubusercontent.com/61015245/195039949-9ee18637-2698-4c07-ad38-ba314e1bf58e.png)

yroeht committed:
mmu: do not log before timer init

This caused a call to arm_arch_timer_count() before
arm_arch_timer_init().

Signed-off-by: Théophile Ranquet <theophile.ranquet@gmail.com>

ibirnbaum commented:
Okay, it's definitely time for some coffee... I don't know where I spotted a LOG_DBG in the aarch32 MMU code earlier on, that code should be OK. It's just that LOG_DBG in kernel/mmu.c

yroeht commented:
I agree.

I am running into some build issues, maybe you can help? `./zephyr/scripts/twister -T zephyr/tests -p qemu_cortex_a9` runs into compilation issues because it does not find my modules (in ./modules).

i.e.

```
west build --pristine -b qemu_cortex_a9 zephyr/tests/crypto/tinycrypt/                                                                                               11:14:29
-- west build: making build dir /home/tranquet/zephyrproject-yroeht/build pristine                                                                                                                                
-- west build: generating a build system                                                                                                                                                                          
Loading Zephyr default modules (Zephyr base).
...
/home/tranquet/zephyrproject-yroeht/zephyr/tests/crypto/tinycrypt/src/hmac.c:40:10: fatal error: tinycrypt/hmac.h: No such file or directory                                                                      
   40 | #include <tinycrypt/hmac.h>
```

This one is solved by `export ZEPHYR_EXTRA_MODULES=~/nobackup/GIT/zephyrproject/modules/crypto/tinycrypt`, but I don't want to do this for every single module. Any solution?

carlocaione commented:
> Maybe we should notify the logging folks and let them know that some flag indicating that logging calls should not yet try to get a timestamp, to be cleared once the timer is ready, should probably be implemented?

This looks to me like a more general issue that we need to address. So, IIUC, the problem is that the logging framework is trying to get a timestamp before the timer is actually initialized? What's the relation here with MMU and syscon?

ibirnbaum commented:
> > Maybe we should notify the logging folks and let them know that some flag indicating that logging calls should not yet try to get a timestamp, to be cleared once the timer is ready, should probably be implemented?
> 
> This looks to me like a more general issue that we need to address. So, IIUC, the problem is that the logging framework is trying to get a timestamp before the timer is actually initialized? What's the relation here with MMU and syscon?

The Syscon driver just happened to be the first driver to be initialized on this particular target that uses the MMU to map its own register space to a virtual address. The syscon's functionality is unrelated to the issue.

The issue is that any device driver using the MMU facilities causes the crash if processing of LOG_DBG statements is enforced as long as the timer isn't initialized due to this one LOG_DBG call at this very central spot in the kernel's MMU code (which as far as I can tell has been around since the early days of this source file). This unfortunately also affects the ARMv7 timer driver itself on systems where the MMU is available: it never gets to the point where it's operational, as the timer driver also attempts to make use of the MMU. It will therefore also run into the LOG_DBG call in the kernel code, and an attempt to read the timer's value will be made by the logging module while trying to initialize the timer. Therefore, this isn't an ordering issue, it's a chicken-or-the-egg problem which likely affects all MMU-based arches in which the system timer is mapped to a virtual address.

As the LOG_DBG statement in the MMU code has been around for quite a while, I'm wondering if this particular logging test has been added only recently? Strange that this issue hasn't popped up earlier.

carlocaione commented:
Can you open a separate PR only with the latest commit explaining carefully why is that needed?

povergoing commented:
> The issue is that any device driver using the MMU facilities causes the crash if processing of LOG_DBG statements is enforced as long as the timer isn't initialized due to this one LOG_DBG call at this very central spot in the kernel's MMU code (which as far as I can tell has been around since the early days of this source file). This unfortunately also affects the ARMv7 timer driver itself on systems where the MMU is available: it never gets to the point where it's operational, as the timer driver also attempts to make use of the MMU. It will therefore also run into the LOG_DBG call in the kernel code, and an attempt to read the timer's value will be made by the logging module while trying to initialize the timer. Therefore, this isn't an ordering issue, it's a chicken-or-the-egg problem which likely affects all MMU-based arches in which the system timer is mapped to a virtual address.

We might need a limited log/printk before MMU and timer are inited. AFAIK, without MMU enabled the uart can also work (actually any device can be simply configured when MMU is disabled I think). At least for v8a/r, when MMU is disabled, the cores have full access to all MMIO regions.

yroeht commented:
> > The issue is that any device driver using the MMU facilities causes the crash if processing of LOG_DBG statements is enforced as long as the timer isn't initialized due to this one LOG_DBG call at this very central spot in the kernel's MMU code (which as far as I can tell has been around since the early days of this source file). This unfortunately also affects the ARMv7 timer driver itself on systems where the MMU is available: it never gets to the point where it's operational, as the timer driver also attempts to make use of the MMU. It will therefore also run into the LOG_DBG call in the kernel code, and an attempt to read the timer's value will be made by the logging module while trying to initialize the timer. Therefore, this isn't an ordering issue, it's a chicken-or-the-egg problem which likely affects all MMU-based arches in which the system timer is mapped to a virtual address.
> 
> We might need a limited log/printk before MMU and timer are inited. AFAIK, without MMU enabled the uart can also work (actually any device can be simply configured when MMU is disabled I think). At least for v8a/r, when MMU is disabled, the cores have full access to all MMIO regions.

You might want to check out https://github.com/zephyrproject-rtos/zephyr/pull/51170

ibirnbaum commented:
> > The issue is that any device driver using the MMU facilities causes the crash if processing of LOG_DBG statements is enforced as long as the timer isn't initialized due to this one LOG_DBG call at this very central spot in the kernel's MMU code (which as far as I can tell has been around since the early days of this source file). This unfortunately also affects the ARMv7 timer driver itself on systems where the MMU is available: it never gets to the point where it's operational, as the timer driver also attempts to make use of the MMU. It will therefore also run into the LOG_DBG call in the kernel code, and an attempt to read the timer's value will be made by the logging module while trying to initialize the timer. Therefore, this isn't an ordering issue, it's a chicken-or-the-egg problem which likely affects all MMU-based arches in which the system timer is mapped to a virtual address.
> 
> We might need a limited log/printk before MMU and timer are inited. AFAIK, without MMU enabled the uart can also work (actually any device can be simply configured when MMU is disabled I think). At least for v8a/r, when MMU is disabled, the cores have full access to all MMIO regions.

When the MMU is disabled, the software has full access to any physical address. However, there's one aspect that on an aarch32 system basically forces you to have MMU support: the hand full of instructions that implement unaligned load/store operations (you'll find those when dissecting Ethernet packets, for example) will crash if the MMU is not set up and enabled.

The UART driver is initialized after the basic MMU setup has been performed, which consists of identity mappings for Zephyr code and data and the exception vectors. The 2nd stage of the base MMU initialization are the mappings that are derived from the SoC the software is running on, Afterwards, the device drivers one by one request their memory mappings. Therefore, the UART driver can't access its registers before its init function has completed. An early printk would therefore only work before the MMU is enabled, and up to that point, the initialization is silent anyways.

ibirnbaum commented:
@yroeht From my point of view, the LOG_DBG call in kernel/mmu.c should remain in place. The issue we've run into here is too limited in scope in order to have a valid argument to remove something from kernel code. #51170 is fine and there's certainly no harm in having this, but the issue should still be communicated to the logging subsys maintainers. After all, I've just come up with a 2nd potential issue: even if the timer safely returned all zeroes as timestamp, with immediate logging enabled, the system would still crash, as the UART driver hasn't mapped its registers at that point either.

I'll open a separate issue report for the maintainers of the logging subsys.

yroeht commented:
> @yroeht From my point of view, the LOG_DBG call in kernel/mmu.c should remain in place.

I agree. The commit that removes LOG_DBG is only present in this PR because it would be a hassle to update it each time https://github.com/zephyrproject-rtos/zephyr/pull/51170 is updated. This PR has to remain on hold until that issue is dealt with.

povergoing commented:
> The UART driver is initialized after the basic MMU setup has been performed, which consists of identity mappings for Zephyr code and data and the exception vectors. The 2nd stage of the base MMU initialization are the mappings that are derived from the SoC the software is running on, Afterwards, the device drivers one by one request their memory mappings. Therefore, the UART driver can't access its registers before its init function has completed. An early printk would therefore only work before the MMU is enabled, and up to that point, the initialization is silent anyways.

What about creating a tmp mmu region for uart in order to do early printk, and the tmp region retired after uart fully enabled?

ibirnbaum commented:
> > The UART driver is initialized after the basic MMU setup has been performed, which consists of identity mappings for Zephyr code and data and the exception vectors. The 2nd stage of the base MMU initialization are the mappings that are derived from the SoC the software is running on, Afterwards, the device drivers one by one request their memory mappings. Therefore, the UART driver can't access its registers before its init function has completed. An early printk would therefore only work before the MMU is enabled, and up to that point, the initialization is silent anyways.
> 
> What about creating a tmp mmu region for uart in order to do early printk, and the tmp region retired after uart fully enabled?

The UART driver as of recently uses the DEVICE_MMIO macros, which automate the mapping of the registers via the MMU and the resolving of the virtual address whenever the registers are accessed. There's no alternative temporary way of access (i.e., via the phyiscal address) contained in this facility. Even if a static PA=VA mapping initially existed right after the MMU is enabled, there would a) be no well-defined means of getting rid of it later on and b) an incompatibility with the memory management API that is encouraged nowadays for use in device drivers.
